netebpfext/net_ebpf_ext_sock_addr.c

```diff
@@ -423,7 +839,479 @@ net_ebpf_extension_sock_addr_authorize_connection_classify(
+        goto Exit;
+    }
+    memset(sock_addr_ctx, 0, sizeof(bpf_sock_addr_t));
+    memset(sock_addr_ctx_original, 0, sizeof(bpf_sock_addr_t));
+
+    _net_ebpf_extension_sock_addr_copy_wfp_connection_fields(incoming_fixed_values, sock_addr_ctx);
+    *sock_addr_ctx_original = *sock_addr_ctx;
+
+    // Check if this call is intended for us.
+    filter_context = (net_ebpf_extension_sock_addr_wfp_filter_context_t*)filter->context;
+    ASSERT(filter_context != NULL);
+    if (filter_context == NULL) {
+        status = STATUS_INVALID_PARAMETER;
+        goto Exit;
+    }
+
+    // If the callout is invoked for v4, then it is safe to simply invoke the eBPF
+    // program from the filter context.
+    // If the callout is invoked for v6:
+    // 1. Check if the destination is v4-mapped v6 address or pure v6 address.
+    // 2. If it is v4-mapped v6 address, then we should procced only if this callout
+    //    is invoked for v4 attach type.
+    // 3. If it is pure v6 address, then we should procced only if this callout is
+    //    invoked for v6 attach type.
+    if (sock_addr_ctx->family == AF_INET6) {
+        if (IN6_IS_ADDR_V4MAPPED((IN6_ADDR*)sock_addr_ctx->user_ip6)) {
+            v4_mapped = true;
+        }
+        if (v4_mapped) {
+            if (!filter_context->v4_attach_type) {
+                // This callout is for v6 attach type, but address is v4-mapped v6 address.
+                // Change action to permit and return.
+                NET_EBPF_EXT_LOG_MESSAGE(
+                    NET_EBPF_EXT_TRACELOG_LEVEL_VERBOSE,
+                    NET_EBPF_EXT_TRACELOG_KEYWORD_SOCK_ADDR,
+                    "net_ebpf_extension_sock_addr_redirect_connection_classify: v6 attach type, v4mapped, ignoring");
+                action = FWP_ACTION_PERMIT;
+                goto Exit;
+            }
+        } else if (filter_context->v4_attach_type) {
+            // This callout is for v4 attach type, but address is a pure v6 address.
+            // Change action to permit and return.
+            NET_EBPF_EXT_LOG_MESSAGE(
+                NET_EBPF_EXT_TRACELOG_LEVEL_VERBOSE,
+                NET_EBPF_EXT_TRACELOG_KEYWORD_SOCK_ADDR,
+                "net_ebpf_extension_sock_addr_redirect_connection_classify: v4 attach type, purev6, ignoring");
+            action = FWP_ACTION_PERMIT;
+            goto Exit;
+        }
+    }
+
+    attached_client = (net_ebpf_extension_hook_client_t*)filter_context->base.client_context;
+    if (attached_client == NULL) {
+        status = STATUS_INVALID_PARAMETER;
+        goto Exit;
+    }
+    if (!net_ebpf_extension_hook_client_enter_rundown(attached_client)) {
+        attached_client = NULL;
+        // Client is detaching, change action to permit.
+        action = FWP_ACTION_PERMIT;
+        goto Exit;
+    }
+
+    // Get the redirect handle for this filter.
+    status = _net_ebpf_ext_sock_addr_get_redirect_handle(filter->filterId, &redirect_handle);
+    if (!NT_SUCCESS(status)) {
```

Comment:
Please make this change in a subsequent PR.

Each function calls RETURN_STATUS macro which says "this function is retuning failure code xyz". Caller of functions will simply call BAIL_ON_ERROR macro. So each function prints error on its return. This reduces code size. Use a specific log macro such as `NET_EBPF_EXT_LOG_MESSAGE_UINT64_UINT64` at the point of failure rather than the caller.